qadence/ml_tools/train_utils/config_manager.py

```diff
@@ -0,0 +1,159 @@
+        """
+        Create a log folder in the specified root folder, adding subfolders if required.
+
+        Args:
+            root_folder (Union[str, Path]): The root folder where the log folder will be created.
+
+        Returns:
+            Path: The path to the created log folder.
+        """
+        root_folder_path = Path(root_folder)
+        root_folder_path.mkdir(parents=True, exist_ok=True)
+
+        if self.config.create_subfolder_per_run:
+            self._add_subfolder()
+            log_folder = root_folder_path / self.config._subfolders[-1]
+        else:
+            if len(self.config._subfolders) == 0:
+                self._add_subfolder()
+            log_folder = root_folder_path / self.config._subfolders[-1]
+
+        log_folder.mkdir(parents=True, exist_ok=True)
+        return Path(log_folder)
+
+    def _add_subfolder(self) -> None:
+        """
+        Add a unique subfolder name to the configuration for logging.
+
+        The subfolder name includes a run ID, timestamp, and process ID in hexadecimal format.
+        """
+        timestamp = datetime.datetime.now().strftime("%Y%m%dT%H%M%S")
+        pid_hex = hex(os.getpid())[2:]
+        run_id = len(self.config._subfolders) + 1
+        subfolder_name = f"{run_id}_{timestamp}_{pid_hex}"
+        self.config._subfolders.append(str(subfolder_name))
+
+    def _handle_hyperparams(self) -> None:
+        """
+        Handle and filter hyperparameters based on the selected tracking tool.
+
+        Removes incompatible hyperparameters when using TensorBoard.
+        """
+        # tensorboard only allows for certain types as hyperparameters
+
+        if (
+            self.config.hyperparams
+            and self.config.tracking_tool == ExperimentTrackingTool.TENSORBOARD
+        ):
+            self._filter_tb_hyperparams()
+
+    def _filter_tb_hyperparams(self) -> None:
+        """
+        Filter out hyperparameters that cannot be logged by TensorBoard.
+
+        Logs a warning for the removed hyperparameters.
+        """
+
+        # tensorboard only allows for certain types as hyperparameters
+        tb_allowed_hyperparams_types: tuple = (int, float, str, bool, Tensor)
+        keys_to_remove = [
+            key
+            for key, value in self.config.hyperparams.items()
+            if not isinstance(value, tb_allowed_hyperparams_types)
+        ]
+        if keys_to_remove:
+            logger.warning(
+                f"Tensorboard cannot log the following hyperparameters: {keys_to_remove}."
+            )
+            for key in keys_to_remove:
+                self.config.hyperparams.pop(key)
+
+    def _derive_parameters(self) -> None:
+        """
+        Derive additional parameters for the training configuration.
+
+        Sets the stopping criterion if it is not already defined.
+        """
+        if self.config.trainstop_criterion is None:
+            self.config.trainstop_criterion = lambda x: x <= self.config.max_iter
```

Comment:
It would be better to keep an open ended nomenclature - in case we would like to derive other parameters based on config in the future.
trainstop_criterion could be only one of the derivations. What do you think?

Reply:
Fair, but the name remains a bit obscure to me. E.g. one could think of `parameters` as the trainable ones. Why not calling it `setup_additional_configuration` or something like that?

Reply:
Updated.